### ISSUE TITLE
fsck: appears totally broken on my vintage hardware

### ISSUE THREAD

**Issue by cradek** [rundel-tech/ZARC#2]
Hi!  I've tried your FSCK.COM (from commit 2584a09) on my Imsai, a Z80 S100 machine running CP/M 2.2 with an 8" single sided single density (standard IBM 3740 format) floppy.  The machine has 32k of well-tested RAM, and a VersaFloppy II interface card.  The drive is the standard 128 bytes/s, 26 s/t, 77 track.  My CP/M bios's sector translation table is the standard/recommended interleave of 6 as used on distributed CP/M floppies:

```
trans   db 1,7,13,19,25,5,11,17
        db 23,3,9,15,21,2,8,14
        db 20,26,6,12,18,24,4
        db 10,16,22
```

Although the system works correctly, and I don't think there's anything wrong with this floppy that I use all the time, the output of fsck shows many errors (perhaps for every file) and it aborts with a Software panic.

When I transfer an image of this floppy to my Debian machine, fsck.cpm and cpmls from the cpmtools package operate on it correctly.

```
chris@green:~$ fsck.cpm imsai-cpm-floppy.img 
Phase 1: check extent fields
Phase 2: check extent connectivity
imsai-cpm-floppy.img: 34/64 files (0.7% non-contigous), 210/243 blocks
chris@green:~$ cpmls -d imsai-cpm-floppy.img 
MOVCPM   COM : PIP      COM : SUBMIT   COM : XSUB     COM
ED       COM : ASM      COM : DDT-ORIG COM : LOAD     COM
STAT     COM : SYSGEN   COM : DUMP     COM : DUMP     ASM
BIOS     ASM : CBIOS    ASM : DEBLOCK  ASM : DISKDEF  LIB
CPM2/20  COM : HW       ASM : ED-ORIG  COM : MBASIC   COM
VF       COM : STATUS   COM : UMPIRE   COM : MONITOR  ASM
DDT      COM : MONITOR  COM : RESOURCE COM : HW       COM
CRC      COM : M80      COM : L80      COM : TRIG     BAS
chris@green:~$
```

I hesitate to attach the floppy image here because I'm unsure I have the right to redistribute everything on it.

Here is a transcript of my FSCK.COM run and a few more things showing files working correctly that it says have errors.

```
A>dir
A: MOVCPM   COM : PIP      COM : SUBMIT   COM : XSUB     COM
A: ED       COM : ASM      COM : DDT-ORIG COM : LOAD     COM
A: STAT     COM : SYSGEN   COM : DUMP     COM : DUMP     ASM
A: BIOS     ASM : CBIOS    ASM : DEBLOCK  ASM : DISKDEF  LIB
A: CPM2/20  COM : HW       ASM : ED-ORIG  COM : MBASIC   COM
A: VF       COM : STATUS   COM : UMPIRE   COM : MONITOR  ASM
A: DDT      COM : MONITOR  COM : RESOURCE COM : HW       COM
A: CRC      COM : M80      COM : L80      COM : TRIG     BAS
A: FSCK     COM
A>fsck
*** CP/M File System Checker V1.1 ***

Operand error. Expected format: fsck <drive>[:] [/s]
Returning to CP/M.

A>fsck a:
*** CP/M File System Checker V1.1 ***

U2 ????????.??? - bad file name
U2 ????????.??? - bad S1
U2 ????????.??? - block number out of range
U2 ????????.??? - bad file name
U2 ????????.??? - bad S1
U2 ????????.??? - duplicated physical extent
U2 ????????.??? - bad file name
U2 ????????.??? - bad S1
U2 ????????.??? - duplicated physical extent
U2 ????????.??? - bad file name
U2 ????????.??? - bad S1
U2 ????????.??? - duplicated physical extent
U0 MOVCPM  .COM - block not unique
U0 PIP     .COM - block number out of range
U0 SUBMIT  .COM - block number out of range
U0 XSUB    .COM - block number out of range
U0 ????????.??? - bad file name
U0 ????????.??? - no allocations in physical extent [warning]
U0 ????????.??? - bad file name
U0 ????????.??? - duplicated physical extent
U0 ????????.??? - bad file name
U0 ????????.??? - duplicated physical extent
U0 ????????.??? - bad file name
U0 ????????.??? - duplicated physical extent
U0 VF      .COM - block number out of range
U0 STATUS  .COM - block number out of range
U0 UMPIRE  .COM - block number out of range
U0 MONITOR .ASM - block number out of range
U61 ???!?v"z.?!w - bad user number
U61 ???!?v"z.?!w - bad file name
U61 ???!?v"z.?!w - bad S1

*** Software panic at PC=06B5 ***

A>user 61
61?

A>user 2
A>dir
No file
A>user 0
A>stat monitor.asm

 Recs  Bytes  Ext Acc
    1     1k    1 R/W A:MONITOR.ASM
Bytes Remaining On A: 29k

A>type monitor.asm
        org     0100H
        jmp     0F000H
        end

A>asm monitor.aax
CP/M ASSEMBLER - VER 2.0


 0100                   org     0100H
 0100 C300F0            jmp     0F000H
 0103                   end
0103
000H USE FACTOR
END OF ASSEMBLY

A>era monitor.hex
A>stat *.*

 Recs  Bytes  Ext Acc
   64     8k    1 R/W A:ASM.COM
   96    12k    1 R/W A:BIOS.ASM
   69     9k    1 R/W A:CBIOS.ASM
   68     9k    1 R/W A:CPM2/20.COM
   24     3k    1 R/W A:CRC.COM
   38     5k    1 R/W A:DDT.COM
   38     5k    1 R/W A:DDT-ORIG.COM
   80    10k    1 R/W A:DEBLOCK.ASM
   49     7k    1 R/W A:DISKDEF.LIB
   33     5k    1 R/W A:DUMP.ASM
    4     1k    1 R/W A:DUMP.COM
   52     7k    1 R/W A:ED.COM
   52     7k    1 R/W A:ED-ORIG.COM
   28     4k    1 R/W A:FSCK.COM
    1     1k    1 R/W A:HW.ASM
    1     1k    1 R/W A:HW.COM
   84    11k    1 R/W A:L80.COM
   14     2k    1 R/W A:LOAD.COM
  158    20k    2 R/W A:M80.COM
  190    24k    2 R/W A:MBASIC.COM
    1     1k    1 R/W A:MONITOR.ASM
    1     1k    1 R/W A:MONITOR.COM
   76    10k    1 R/W A:MOVCPM.COM
   58     8k    1 R/W A:PIP.COM
   84    11k    1 R/W A:RESOURCE.COM
   41     6k    1 R/W A:STAT.COM
   12     2k    1 R/W A:STATUS.COM
   10     2k    1 R/W A:SUBMIT.COM
    8     1k    1 R/W A:SYSGEN.COM
    2     1k    1 R/W A:TRIG.BAS
   12     2k    1 R/W A:UMPIRE.COM
  116    15k    1 R/W A:VF.COM
    6     1k    1 R/W A:XSUB.COM
Bytes Remaining On A: 29k

A>dump fsck.com

0000 ED 73 A2 0E 31 11 0F CD 72 0B D2 18 01 11 4B 08
0010 0E 09 CD 05 00 C3 00 00 21 21 08 CD BE 0B 21 81
0020 00 CD CE 0B 7E 23 D6 41 38 46 FE 10 30 42 32 A4
0030 0E 7E FE 3A 20 01 23 AF 32 A5 0E CD CE 0B 7E 23
0040 4F A7 28 32 FE 5B 28 08 FE 2F 28 04 FE 24 20 20
(I cut this off here)
```

Please let me know how I can help.

**Comment by rundel-tech** [comment]
Hi Cradek.

Thank you for trying this, and for your very helpful bug report. I think your suspicions are correct, in that FSCK does not handle sector translation correctly (or at all)! I don't see the issue on my system as it uses a FLASH memory card and hence no sector translation. Unfortunately, my hardware is in storage for the time being but I will attempt to investigate this rather serious bug regardless. I'll keep you posted.

Regards,
    Merlin

**Comment by cradek** [comment]
Hi Merlin, thanks for your response.  In pull request #3 I added sector translation, and it should not break on your system if your BIOS implements a null SECTRAN.  However there must still be missing pieces. I'm not sure what to do about the remaining problem.  I will keep studying.

Cheers,
Chris

**Comment by cradek** [comment]
Full output with PR #3:

```
A>save 14 fsck2.com
A>fsck2 a:
*** CP/M File System Checker V1.1 ***

U0 MOVCPM  .COM - block number out of range
U0 PIP     .COM - block number out of range
U0 SUBMIT  .COM - block number out of range
U0 XSUB    .COM - block number out of range
U0 ED      .COM - block number out of range
U0 ASM     .COM - block number out of range
U0 DDT-ORIG.COM - block number out of range
U0 LOAD    .COM - block number out of range
U0 STAT    .COM - block number out of range
U0 SYSGEN  .COM - block number out of range
U0 DUMP    .COM - block number out of range
U0 DUMP    .ASM - block number out of range
U0 BIOS    .ASM - block number out of range
U0 CBIOS   .ASM - block number out of range
U0 DEBLOCK .ASM - block number out of range
U0 DISKDEF .LIB - block number out of range
U0 CPM2/20 .COM - block number out of range
U0 HW      .ASM - block number out of range
U0 ED-ORIG .COM - block number out of range
U0 MBASIC  .COM - block number out of range
U0 VF      .COM - block number out of range
U0 STATUS  .COM - block number out of range
U0 UMPIRE  .COM - block number out of range
U0 MONITOR .ASM - block number out of range
U0 DDT     .COM - block number out of range
U0 MONITOR .COM - block number out of range
U0 RESOURCE.COM - block number out of range
U0 HW      .COM - block number out of range
U0 CRC     .COM - block number out of range
U0 MBASIC  .COM - block number out of range
U0 M80     .COM - block number out of range
U0 M80     .COM - block number out of range
U0 L80     .COM - block number out of range
U0 TRIG    .BAS - block number out of range
U0 FSCK    .COM - block number out of range
U0 FSCK2   .COM - block number out of range
Directory checks complete
Errors and warnings: 36
Files found: 34
Unused directory entries: 28
Blocks used: 39

Block map ('D': directory, 'F': file and '-': unused).
DDF---------F-------F-FF------F-------F----F-F-----FFF----F-----------F--------F
---------F------F--------FF-F-----F----F---FF-----------FFFF---F----------F-FF--
-------------F-------F---------------F---F-------------F------------------------
---
Surface scan skipped
Returning to CP/M.

```

**Comment by cradek** [comment]
Here is a dump of T2 S17 showing that double M80.COM (shown twice in the above output), but which is just one file, with two extents

```
4800 0043 5243 2020 2020 2043 4F4D 0000 0018 |.CRC     COM....|
4810 9A9B 9C00 0000 0000 0000 0000 0000 0000 |................|
4820 004D 4241 5349 4320 2043 4F4D 0100 003E |.MBASIC  COM...>|
4830 ADAE AFB0 B1B2 B3B4 0000 0000 0000 0000 |................|
4840 004D 3830 2020 2020 2043 4F4D 0000 0080 |.M80     COM....|
4850 B5B6 B7B8 B9BA BBBC BDBE BFC0 C1C2 C3C4 |................|
4860 004D 3830 2020 2020 2043 4F4D 0100 001E |.M80     COM....|
4870 C5C6 C7C8 0000 0000 0000 0000 0000 0000 |................|
```

```
A>stat m80.com

 Recs  Bytes  Ext Acc
  158    20k    2 R/W A:M80.COM
Bytes Remaining On A: 25k

```

**Comment by rundel-tech** [comment]
I may have found it. fsck.z80 line 567 currently reads:
               jr z, slf_wd_alu_dn     ; Jump if no unused entries encountered
Should read:
               jr z, slf_by_alu_dn     ; Jump if no unused entries encountered
Would you mind trying this? It looks promising!

**Comment by cradek** [comment]
Yes!  With my PR #3 and then your change, I get a much more plausible output:

```
A>fsck2 a:
*** CP/M File System Checker V1.1 ***

U0 MBASIC  .COM - no allocations in physical extent [warning]
U0 M80     .COM - no allocations in physical extent [warning]
Directory checks complete
Errors and warnings: 2
Files found: 34
Unused directory entries: 28
Blocks used: 218

Block map ('D': directory, 'F': file and '-': unused).
DDFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF----------------------
---
Surface scan skipped
Returning to CP/M.
```

However I think both those remaining warnings are wrong: you can see in the paste above https://github.com/rundel-tech/ZARC/issues/2#issuecomment-1589841789 that both those files have second extents (File extent byte = 01) that are nonempty.  Those two files are the only ones that use more than one extent.

**Comment by cradek** [comment]
I updated PR #3 accordingly

**Comment by rundel-tech** [comment]
I think I have found the remaining issue. When scanning blocks in each extent, the "C" register is used to count the number of blocks found plus some flags. Unfortunately, only four bits are allocated to the counter. This is sufficient to count eight 16-bit entries, but overflows to zero when faced with a fully allocated extent with 16 8-bit entries. The fix: change fsck.i line 25, currently:

ALLCF_COUNT     equ 0x0f                ; Used allocation counter

should be:

ALLCF_COUNT     equ 0x1f                ; Used allocation counter

If this works and you are happy with it, I will merge the changes and make bump the version.

**Comment by cradek** [comment]
Again you found it; I have correct output now.  I added this change to pull request #3.

```
A>save 14 fsck2.com
A>fsck2 a:
*** CP/M File System Checker V1.1 ***

Directory checks complete
Errors and warnings: 0
Files found: 34
Unused directory entries: 28
Blocks used: 218

Block map ('D': directory, 'F': file and '-': unused).
DDFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF-F---------------------
---
Surface scan skipped
Returning to CP/M.

A>fsck2 a: /s
*** CP/M File System Checker V1.1 ***

Directory checks complete
Errors and warnings: 0
Files found: 34
Unused directory entries: 28
Blocks used: 218

Block map ('D': directory, 'F': file and '-': unused).
DDFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF-F---------------------
---

Checking all blocks are readable (surface scan)
Track: 76
Scan complete
Returning to CP/M.

A>

```

**Comment by cradek** [comment]
In my opinion, these bug fixes are complete and PR #3 is ready for merge.  Thank you for your help!

**Comment by rundel-tech** [comment]
I have completed the merge and updated the revision to V1.2. Many thanks for your help. I really appreciate that FSCK has been tested on period hardware and for your help in debugging. Feel free to try bitcoin mining (https://www.tech-cave.co.uk/2022/11/bitcoin-mining-on-z80.html) if you fancy it!
